hasura/ddn-docs
# Bug Bash: Fix Safari copy issue

## Description

This PR resolves the issue with copying on Safari. Turns out, a Docusaurus dependency has an [unpatched bug](https://github.com/sindresorhus/copy-text-to-clipboard/issues/47). This closes hasura/graphql-engine#10341 and closes hasura/graphql-engine#10349.

[DOCS-2205](https://hasurahq.atlassian.net/browse/DOCS-2205)

## Design

This PR swizzles the existing `<CopyButton />` component of Docusuarus. Honestly, little code is changed, but we've swapped out the dependency on [the `copy-text-to-clipboard` package](https://github.com/sindresorhus/copy-text-to-clipboard) to instead use [the `navigator.clipboard` API](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/clipboard), which enjoys support across all modern browsers. Why is Docusaurus using this dep? No clue.

We have a utility function that will check to ensure the API is available and then copy the passed string:

```ts
export function copyToClipboard(text: string): Promise<void> {
  // First, we'll check to make sure the clipboard API is available. If it is, copy that text.
  if (navigator.clipboard && navigator.clipboard.writeText) {
    return navigator.clipboard.writeText(text);
  } else {
    // Just in case navigator.clipboard isn't supported...which is very unlikely.
    console.log('This is a very old version of a browser.');
  }
}
```

Which is then used in the component via a callback:

```tsx
const handleCopyCode = useCallback(async () => {
    try {
      await copyToClipboard(code);
      setIsCopied(true);
      copyTimeout.current = window.setTimeout(() => {
        setIsCopied(false);
      }, 1000);
    } catch (err) {
      console.error('Failed to copy text:', err);
      setIsCopied(false);
    }
  }, [code]);
```

☝️ Much of the original design here is borrowed from the ejected component via Docusaurus.

## Testing

1. Open windows for Safari, Chrome, Firefox (and any other browser you wish).
2. Navigate to [this link](https://rob-bug-bash-fix-safari-copy.v3-docs-eny.pages.dev/getting-started/build/connect-to-data/connect-a-source?db=PostgreSQL) in each.
3. Attempt to copy multiple code blocks / snippets from within each browser.
4. Paste them to ensure behavior is as expected 🤞 


[DOCS-2205]: https://hasurahq.atlassian.net/browse/DOCS-2205?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
